[SPARK-15083] [Web UI] History Server can OOM due to unlimited TaskUIData

ajbozarth opened the pull request:
## What changes were proposed in this pull request?

Based on #12990 by @tankkyo

Since the History Server currently loads all application's data it can OOM if too many applications have a significant task count. `spark.ui.trimTasks` (default: false) can be set to true to trim tasks by `spark.ui.retainedTasks` (default: 10000) 

(This is a "quick fix" to help those running into the problem until a update of how the history server loads app data can be done)
## How was this patch tested?

Manual testing and dev/run-tests

![spark-15083](https://cloud.githubusercontent.com/assets/13952758/17713694/fe82d246-63b0-11e6-9697-b87ea75ff4ef.png)

ajbozarth committed:
Added optonal task trimming (based on @tankkyo code)

SparkQA commented:
**[Test build #63874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63874/consoleFull)** for PR 14673 at commit [`d0aeae6`](https://github.com/apache/spark/commit/d0aeae6d8c9b623a2f2966af8803c3d7933f6c51).

ajbozarth commented:
Found a odd behavior while testing. When a application encounters an issue (in my case my task queue filled up and SparkListenerBus was stopped) it can cause the tasks HashMap size to be larger than the sum of completed, failed, and active tasks (total tasks). This means that the UI can end up looking like the following two screen shots (of the same application). If everyone can agree this is a separate issue from this I will file a JIRA for it for later.

with `spark.ui.trimTasks` true:
![screen shot 2016-08-16 at 2 29 07 pm](https://cloud.githubusercontent.com/assets/13952758/17716514/d2dc7d88-63bd-11e6-8821-0cbcc41033b8.png)

with `spark.ui.trimTasks` false:
![screen shot 2016-08-16 at 2 20 40 pm](https://cloud.githubusercontent.com/assets/13952758/17716496/c522ecf4-63bd-11e6-8bd1-a922f9fa00b4.png)

ajbozarth committed:
edited text to be less confusing in case of application error

ajbozarth commented:
Pushed a update to the text so it's clearer in case of the above bug:
![screen shot 2016-08-16 at 2 47 28 pm](https://cloud.githubusercontent.com/assets/13952758/17717030/665bc92c-63c0-11e6-97e5-a4a83b307965.png)
![screen shot 2016-08-16 at 2 48 30 pm](https://cloud.githubusercontent.com/assets/13952758/17717055/8cd0f816-63c0-11e6-9778-2bcb5a9d1247.png)

ajbozarth commented:
@vanzin and @tgravescs here's my update of pr #12990 by @tankkyo if you wan to take a look, I raised a few discussion points above

SparkQA commented:
**[Test build #63876 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63876/consoleFull)** for PR 14673 at commit [`f67a321`](https://github.com/apache/spark/commit/f67a3215740660859be2947d0dc17f08eb5ffbe1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63876/
Test FAILed.

SparkQA commented:
**[Test build #63874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63874/consoleFull)** for PR 14673 at commit [`d0aeae6`](https://github.com/apache/spark/commit/d0aeae6d8c9b623a2f2966af8803c3d7933f6c51).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63874/
Test PASSed.

tgravescs commented:
yes the queue filling up and dropping events is a separate issue there are other jiras filed for it.

tgravescs commented on core/src/main/scala/org/apache/spark/ui/SparkUI.scala:
```diff
@@ -141,6 +141,7 @@ private[spark] object SparkUI {
   val DEFAULT_POOL_NAME = "default"
   val DEFAULT_RETAINED_STAGES = 1000
   val DEFAULT_RETAINED_JOBS = 1000
+  val DEFAULT_RETAINED_TASKS = 10000
```
do we have an idea of when people are hitting this?  Personally I find 10000 small also, I was thinking more like 50000 or 100000 and users can limit further if they have the history server configured to use small amounts of memory.

ajbozarth replied:
I'll change it to 100000

ajbozarth replied:
As for the why, the issue is when the history server is full of applications with large task amounts, there is no issue with just one application having a significant number of tasks by itself.

tgravescs commented on core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala:
```diff
@@ -93,6 +93,8 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
   val retainedStages = conf.getInt("spark.ui.retainedStages", SparkUI.DEFAULT_RETAINED_STAGES)
   val retainedJobs = conf.getInt("spark.ui.retainedJobs", SparkUI.DEFAULT_RETAINED_JOBS)
+  val retainedTasks = conf.getInt("spark.ui.retainedTasks", SparkUI.DEFAULT_RETAINED_TASKS)
+  val trimTasks = conf.getBoolean("spark.ui.trimTasks", false)
```
I don't think we need the on/off switch for the trimTasks.  None of the others have off switches so I say just leave it without for now and a user can set to something huge like 100,000,000.  We can always add an off later if needed (like set to -1). If we are worried about an int being to small we could make it a long.  I don't see anything have over 2 billions tasks at this point though.

ajbozarth replied:
Ok, that works for me

tgravescs commented on core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala:
```diff
@@ -93,6 +93,8 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
 
   val retainedStages = conf.getInt("spark.ui.retainedStages", SparkUI.DEFAULT_RETAINED_STAGES)
   val retainedJobs = conf.getInt("spark.ui.retainedJobs", SparkUI.DEFAULT_RETAINED_JOBS)
+  val retainedTasks = conf.getInt("spark.ui.retainedTasks", SparkUI.DEFAULT_RETAINED_TASKS)
```
ideally we use the new config builder stuff

tgravescs replied:
actually looking at the taskId's they are a long so we should make this a long.

ajbozarth replied:
I'm not sure I understand why this should be a long, also I haven't used the new config builder yet, should I update all of the vals to use it?

tgravescs replied:
a long would just allow it to be set to a higher value.  2^64 vs 2^32.  Since the TaskId is a long already (see TaskInfo.scala) it makes sense to make the config a long. It really only matters if someone wants to set the limit over 2 billionish.

We have been trying to make any new configs the ConfigEntry type. so you don't have to do them all just the new one.  Its pretty straight forward, take a look at https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/config/package.scala

ajbozarth replied:
`retainedTasks` is compared to `stageData.taskData.size`, which is a `int`, I don't see how `TaskId` being a long is relevant. 

And I've switched to a `ConfigEntry` and will push my changes after testing.

tgravescs replied:
ah good point that hashMap can only hold int number of entries.  

My point was that with taskId being a long and assuming taskId = 0 through X, X here could be a long (2^64) and thus you could have 2^64 number of tasks.  If you can have 2^64 number of tasks then this config should allow you to configure it to retain all of those tasks.

I'm fine with leaving an int.

ajbozarth replied:
Ah, now I get where you where going with that, yeah if it wasn't for the map size issue that would have been good.

tgravescs commented on core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala:
```diff
@@ -137,6 +139,17 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
+  private def trimTasksIfNecessary(taskData: HashMap[Long, TaskUIData]) = synchronized {
+    if (taskData.size > retainedTasks) {
+      val toRemove = (taskData.size - retainedTasks)
+      val oldIds = taskData.map(_._2.taskInfo.taskId).toList.sorted.take(toRemove)
```
is there a reason we are using _._2.taskInfo.taskId instead of just the Long ( _._1), which is the taskId also?

I'm a bit concerned with the sorting here.  If someone is using really large number of tasks , say 1 million I'm afraid this sorting might be expensive, especially if we are doing it for every single task that comes in after hitting the limit, we have to keep resorting every time.

ajbozarth replied:
This code is from the original pr, I didn't know _._1 is always equal to _._2.taskInfo.taskId so I'll change that, as for the loop, it does slow things down, I noticed it in my testing, I'll look at a better way to do this.

ajbozarth committed:
updated tasks hash so trimming is simplier and simlified confs

ajbozarth commented:
@tgravescs I've addressed some issues: I changed the default max to 100000 and removed the on/off. I also updated the `taskData` to be a `LinkedHashMap` instead of a `HashMap` which allows for easier trimming without losing or changing the current functionality. I'll be pushing the code later today once I finish my tests.

ajbozarth commented:
ran into a odd test failure in HistoryServer I still haven't figured out, I'll push an update to fix it tomorrow

SparkQA commented:
**[Test build #63971 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63971/consoleFull)** for PR 14673 at commit [`014db4c`](https://github.com/apache/spark/commit/014db4c88b6b8a56a8a7a8197c27ac0b6e02f1a9).

SparkQA commented:
**[Test build #63971 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63971/consoleFull)** for PR 14673 at commit [`014db4c`](https://github.com/apache/spark/commit/014db4c88b6b8a56a8a7a8197c27ac0b6e02f1a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63971/
Test FAILed.

ajbozarth committed:
updated config impl and fixed test cases

ajbozarth commented:
Pushed my latest updates including the change to using `ConfigEntry` and updating the test cases (using `LinkedHashMap` changes the secondary sort to `taskId`)

SparkQA commented:
**[Test build #64022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64022/consoleFull)** for PR 14673 at commit [`1faea5f`](https://github.com/apache/spark/commit/1faea5f8390995a6d5f2c62427fde384432a95d6).

SparkQA commented:
**[Test build #64022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64022/consoleFull)** for PR 14673 at commit [`1faea5f`](https://github.com/apache/spark/commit/1faea5f8390995a6d5f2c62427fde384432a95d6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64022/
Test PASSed.

tgravescs commented:
code changes look good, does it change the secondary sort of taskId in a bad way?  Looking at your json files it looks like they were in order before and now they aren't.

ajbozarth commented:
secondary sort in this case is whatever order the collection is stored in, since `LinkedHashMap` keeps an order (essentially by `tasked`, but really by insertion order) `tasked` becomes the defacto secondary sort. In `HashMap` there's no guaranteed order and it's secondary sort is whatever order the hashes are stored (this can be consistent in controlled environments like test suites). 

tl;dr it doesn't change secondary sort in a bad way, it just provides an order that previously didn't exist.

ajbozarth commented:
@srowen @vanzin or @JoshRosen could one of you take a look at this? I've already addressed @tgravescs feedback, and would love to see this go through soon if it looks good to you, thanks

tgravescs commented:
sorry @ajbozarth somehow my mail didn't show me you updated it.

tgravescs commented:
+1, I'll merge shortly

tgravescs commented:
I merged this into 2.1.0, if you are wanting this in older ones can you put up PR for those, cherry-pick to branch-2.0 wasn't clean.

ajbozarth commented:
@tgravescs would would I go about opening a pr for older branches? I've only done work on the latest code before. Also is the 1.6.x branch still updating or is 2.0.x and 2.1 the only updating branches now?

tgravescs commented:
when you do the pr to apache/spark just pick branch-2.0 instead of master.  I see things going into branch-1.6 also but I haven't seen any discussion on the list about doing another release so we would have to ask there.

ajbozarth commented:
ok, thanks, I'll open those pr's once I can cleanly merge the patch

srowen commented:
You would also have to check out branch-2.0 and branch from there in order to work off the right base. From there you can cherry-pick your commit from master and resolve conflicts. Then there's nothing extra to do when opening a PR as it's already known to be vs branch-2.0. It's more than just choosing a branch when you open the PR.

srowen commented:
Ah forgot to comment on the substance: normally we probably wouldn't back-port a feature to a maintenance branch. This is pretty much a feature. If it's low-risk and someone wants to make the case that it is important enough, it's possible. For example, you could argue that the potential for an OOM is a bug. My gut says it would be reasonable for 2.0.x

ajbozarth commented:
Thanks @srowen based on that I'll just open it to the 2.0 branch then
